paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,518 @@
+          output->mutable_data<int8_t>(ctx.GetPlace(), buffer_size);
+      output->set_format((memory::format)dst_prim_desc.desc().data.format);
+
+      return memory(dst_prim_desc, to_void_cast<int8_t>(output_data));
+    }
+  }
+};
+
+static std::string GetHash(const Tensor *input_x, const Tensor *input_y,
+                           const std::string &suffix) {
+  auto dim2str = [](const DDim &operand_dims) {
+    std::string str = "";
+    for (int i = 0; i < operand_dims.size(); ++i) {
+      str += std::to_string(operand_dims[i]) + "-";
+    }
+    return str;
+  };
+
+  std::string hash = std::to_string((unsigned)input_x->format()) +
+                     std::to_string((unsigned)input_x->type()) +
+                     dim2str(input_x->dims()) +
+                     std::to_string((unsigned)input_y->format()) +
+                     std::to_string((unsigned)input_y->type()) +
+                     dim2str(input_y->dims()) + suffix;
+
+  return hash;
+}
+
+template <typename T, typename K>
+std::shared_ptr<MulPrimitiveFactory<T, K>> GetPrimitiveFactory(
+    const MKLDNNDeviceContext &dev_ctx, const ExecutionContext &ctx,
+    const Tensor *input_x, const Tensor *input_y,
+    const mkldnn::engine &mkldnn_engine) {
+  const std::string key = GetHash(input_x, input_y, ctx.op().Output("Out"));
+
+  auto prim_creator =
+      std::static_pointer_cast<MulPrimitiveFactory<T, K>>(dev_ctx.GetBlob(key));
+  if (prim_creator == nullptr) {
+    bool enable_quant =
+        std::is_same<T, int8_t>::value || std::is_same<T, uint8_t>::value;
+    if (enable_quant) {
+      prim_creator =
+          std::make_shared<QuantMulPrimitiveFactory<T, K>>(mkldnn_engine);
+    } else {
+      prim_creator = std::make_shared<MulPrimitiveFactory<T, K>>(mkldnn_engine);
+    }
+    dev_ctx.SetBlob(key, prim_creator);
+  }
+
+  return prim_creator;
+}
+
+template <typename T, typename K>
+class MulMKLDNNKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const ExecutionContext &ctx) const override {
+    PADDLE_ENFORCE(platform::is_cpu_place(ctx.GetPlace()),
+                   "It must use CPUPlace.");
+
+    auto &dev_ctx = ctx.template device_context<MKLDNNDeviceContext>();
+    const auto &mkldnn_engine = dev_ctx.GetEngine();
+
+    const Tensor *x = ctx.Input<Tensor>("X");
+    const Tensor *y = ctx.Input<Tensor>("Y");
+    Tensor *out = ctx.Output<Tensor>("Out");
+    auto out_dims = out->dims();
+
+    auto prim_creator =
+        GetPrimitiveFactory<T, K>(dev_ctx, ctx, x, y, mkldnn_engine);
+    auto mul = prim_creator->CreateMulPrimitive(x, y, out, ctx);
+
+    stream(stream::kind::eager).submit({mul}).wait();
+
+    if (out_dims.size() != 2) {
+      out->Resize(out_dims);
+    }
+    out->set_layout(DataLayout::kMKLDNN);
+    out->set_format(out->format());
+  }
+};
+
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+REGISTER_OP_KERNEL_WITH_CUSTOM_TYPE(mul, MKLDNN, ::paddle::platform::CPUPlace,
```

Comment:
I think it'd be worth to add op registration for plain `fp32` as well.

Reply:
yeah, I think so if the MUL Op wouldn't be used as regular MUL OP, but for now, fp32 x fp32 hasn't been tested in models, so we plan to register plain fp32 in future when it's necessary.